Access C Struct as Numpy Structured Array (#3661)
numba/numba: docs/source/user/cfunc.rst

:func:`numba.farray` should be used instead.

> Dealing with C structures

Reviewer: ```suggestion
Handling C structures
```

> =========================

Reviewer: This adjusts the length to match switching the above to `Handling`.
```suggestion
=====================
```

With CFFI
---------

> For applications that has a lot of states, it is useful to pass data in C

Reviewer: ```suggestion
For applications that have a lot of state, it is useful to pass data in C
```

> structures.  To simplify the interop with C code, numba can convert a ``cffi``

Reviewer: s/interop/interoperability/ (can't use the suggest edit thing here as it'd run past 80 chars)

Author: fixed

type into a numba ``Record`` type using ``numba.cffi_support.map_type``::

   from numba import cffi_support

   nbtype = cffi_support.map_type(cffi_type, use_record_dtype=True)

> .. note:: **use_record_dtype=True** is needed.  Otherwise, pointers to C

Reviewer: ```suggestion
.. note:: **use_record_dtype=True** is needed otherwise pointers to C
```

    structures are returned as void pointers.

For example::

   from cffi import FFI

   src = """

   /* Define the C struct */
   typedef struct my_struct {
      int    i1;
      float  f2;
      double d3;
   } my_struct;

   /* Define a callback function */
   typedef double (*my_func)(my_struct*, size_t);
   """

   ffi = FFI()
   ffi.cdef(src)

>    # Get function signature from *my_func*

Reviewer: ```suggestion
   # Get the function signature from *my_func*
```

   sig = cffi_support.map_type(ffi.typeof('my_func'), use_record_dtype=True)

>    # Make cfunc

Reviewer: ```suggestion
   # Make the cfunc
```

   from numba import cfunc, carray

   @cfunc(sig)
   def foo(ptr, n):
      base = carray(ptr, n)  # view pointer as an array of my_struct
      tmp = 0
      for i in range(n):
         tmp += base[i].i1 * base[i].f2 / base[i].d3
      return tmp

With ``numba.types.Record.make_c_struct``
-----------------------------------------

> The ``numba.types.Record`` type can be created manually to follow the

Reviewer: ```suggestion
The ``numba.types.Record`` type can be created manually to follow a
```

> C-structure layout.  To do that, use ``Record.make_c_struct``::

Reviewer: ```suggestion
C-structure's layout.  To do that, use ``Record.make_c_struct``, for example::
```

   my_struct = types.Record.make_c_struct([
      # Provides a sequence of 2-tuples i.e. (name:str, type:Type)
      ('i1', types.int32),
      ('f2', types.float32),
      ('d3', types.float64),
   ])

> Due to ABI limitations, structures should be passed as pointers.

Reviewer: ```suggestion
Due to ABI limitations, structures should be passed as pointers
```

Reviewer: Need to resolve.

Author: fixed.

> Use ``types.CPointer(my_struct)`` as the argument type.  Inside the ``cfunc``

Reviewer: change to (can't use edit thing, 80 char limit exceeded):
```
using the ``types.CPointer(my_struct)`` as the argument type.  Inside the ``cfunc``
```
